teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -112,9 +112,186 @@ be handled, or user scenarios that will be affected and must be accomodated.
 the Run spec. Validation of the custom task is delegated to the custom controller.
 
 ## Proposal
-TBD
+
+Add support for `Run.RunSpec.Spec`. 
+
+Currently, `Run.RunSpec.Spec` is not supported and there are validations across the
+codebase to ensure, only `Run.RunSpec.Ref` is specified. As part of this TEP, in addition
+to adding support for `Run.RunSpec.Spec`the validations will be changed to support 
+"One of `Run.RunSpec.Spec` or`Run.RunSpec.Ref`" only and not both as part of a single
+API request to kubernetes.
+
+Structure of `RunSpec` after adding the field `Spec` of type `*v1beta1.EmbeddedTask`,
+
+```go
+// RunSpec defines the desired state of Run
+type RunSpec struct {
+	// +optional
+	Ref *TaskRef `json:"ref,omitempty"`
+
+	// TaskSpec is a specification of a custom task
+	// +optional
+	Spec *v1beta1.EmbeddedTask `json:"spec,omitempty"`
+
+	// +optional
+	Params []v1beta1.Param `json:"params,omitempty"`
+
+	// Used for cancelling a run (and maybe more later on)
+	// +optional
+	Status RunSpecStatus `json:"status,omitempty"`
+
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName"`
+
+	// PodTemplate holds pod specific configuration
+	// +optional
+	PodTemplate *PodTemplate `json:"podTemplate,omitempty"`
+
+	// Workspaces is a list of WorkspaceBindings from volumes to workspaces.
+	// +optional
+	Workspaces []v1beta1.WorkspaceBinding `json:"workspaces,omitempty"`
+}
+```
+
+An embedded task will accept a new field i.e. `Spec` with type
+[`runtime.RawExtension`](https://github.com/kubernetes/apimachinery/blob/v0.21.0/pkg/runtime/types.go#L94)
+in addition to `ApiVersion` and `Kind` fields of type string (as part of 
+[`runtime.TypeMeta`](https://github.com/kubernetes/apimachinery/blob/v0.21.0/pkg/runtime/types.go#L36)) :
+
+```go
+type EmbeddedTask struct {
+	// +optional
+	runtime.TypeMeta `json:",inline,omitempty"`
+
+	// +optional
+	Spec runtime.RawExtension `json:"spec,omitempty"`
+
+	// +optional
+	Metadata PipelineTaskMetadata `json:"metadata,omitempty"`
+
+	// TaskSpec is a specification of a task
+	TaskSpec `json:",inline,omitempty"`
+}
```

Comment:
Thank you for adding these details. This is how the new `struct` would look like. 

This is what we have today:

https://github.com/tektoncd/pipeline/blob/3dc52df0d5571b7839929d49ae23e5bb6a11f101/pkg/apis/pipeline/v1beta1/pipeline_types.go#L111-L117